Us 71 notifications

darkheros12 opened the pull request:
<!--
NOTE: Please set, 
- reviewers for the PR
- assign yourself in "Assignees"
- "pull request" label for "Labels"
- set the correct "Milestone",
- set the "Linked Issues" to the issue this PR would close, if any

in the menu on the right
-->

### Related Issues
#71 

### Description
I refactored the NotificationListenerManager (originally AuthFailureNotificationListener).
I added notification functions in SignUpViewModel.
I added a few functions in smsInvitationActivity.
I modified the notification so it return a string not an enum.
I made a "generic" notification channel and builder in NotificationListenerManager so it can adapt to most notification if not all.


### How to check PR
To test:
1. Login as a user.
2. add user by phone number (the user must not exist in the db)
3. Once it is done, log out and registered a new user with the SAME PHONE NUMBER that it was sent previously.
4. Once registered, a notification will pop up for all user that send an invitation.
5. A notification should pop up 

if there is any questions, please comment. THANK YOU

PS: I cannot request @refatK because he help me with this feature. We made a lot of change.

darkheros12 committed:
Create GeneralNotificationListener.java

darkheros12 committed:
Delete GeneralNotificationListener.java

darkheros12 committed:
Merge branch 'staging' into US-71-Notifications

darkheros12 committed:
Merge branch 'staging' into US-71-Notifications

darkheros12 committed:
Initial commit

-Added SmsInvitationRequest

-Renaming AuthFailureNotificationListener to NotificationListenerManager

darkheros12 committed:
full set up for notification

darkheros12 committed:
Feature completed

darkheros12 committed:
SonarCloud fixed

darkheros12 committed:
removed obselete deprecation library

darkheros12 committed:
Merge branch 'staging' into US-71-Notifications

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/viewModel/SignUpViewModel.java:
```diff
@@ -93,4 +101,33 @@ public String randomNumberStringGenerator() {
+
+    public void createNotificationObject(String senderUid, String phoneNumber){
+
+        Notification newNotification = new SmsInvitationRequest();
```
This should be a Notification instead of an SmsInvitationRequest.

darkheros12 replied:
Done.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/model/Notification.java:
```diff
@@ -115,6 +130,13 @@ public long timeSinceCreationMillis() {
         return Timestamp.now().toDate().getTime() - this.getCreationTime().toDate().getTime();
     }
 
+    public String getTitle(){
```
Should have the `@Exclude` annotation since we don't want to push this variable in the database.

darkheros12 replied:
Done.

ShawnGuerra-Bautista reviewed (changes requested):
The feature works, and great job.

However, there are some changes I request.

darkheros12 committed:
PR review fixes

- Added an exclude tag to not affect the database
- change the smsInvitationRequest to Notification because we are creating a notification object.

ShawnGuerra-Bautista reviewed (dismissed):
Good job.

Everything is working fine.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/repository/SmsInvitationRepository.java:
```diff
@@ -0,0 +1,11 @@
+import com.example.bigowlapp.model.SmsInvitationRequest;
+
+public class SmsInvitationRepository extends Repository<SmsInvitationRequest> {
+    public static final String COLLECTION_NAME = "SmsInvitations";
```
The collection name should start with a lower case letter.

ShawnGuerra-Bautista reviewed (changes requested):
Sorry about not noticing this at first, but there's one last change I request.

refatK commented on app/src/main/java/com/example/bigowlapp/activity/LoginPageActivity.java:
```diff
@@ -25,7 +25,7 @@
     private FirebaseAuth.AuthStateListener mAuthStateListener;
     private LogInViewModel logInViewModel;
-    private AuthFailureNotificationListener authListener;
+    private NotificationListenerManager authListener;
```
should also change the variable name itself, still called `authListener`

darkheros12 replied:
done.

refatK commented on app/src/main/java/com/example/bigowlapp/adapter/NotificationAdapter.java:
```diff
@@ -32,7 +32,8 @@ public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
         Notification notification = mNotificationTitles.get(position);
-        holder.getNameTextView().setText(notification.getType().toString());
+        holder.getNameTextView().setText(notification.getTitle());
+        holder.getContentTextView().setText(notification.getMessage() != null ? notification.getMessage() : "");
```
good null check

refatK commented on app/src/main/java/com/example/bigowlapp/utils/NotificationListenerManager.java:
```diff
@@ -37,6 +45,10 @@ public void listen(Context context) {
                         if (notification.getType() == Notification.Type.AUTH_BY_PHONE_NUMBER_FAILURE) {
                             handleAuthByPhoneNumberFailure((AuthByPhoneNumberFailure) notification);
                         }
+
+                        if(notification.getType() == Notification.Type.SMS_INVITATION_REQUEST){
+                            handleSmsInvitationRequest((SmsInvitationRequest) notification, context);
+                        }
```
I think you should just make this an if else statement (rather than 2 if statements) and  make handleSmsInvitationRequest into something that takes all notifs.

This would allow all notifs to have a android notif, and its a pretty simple change. Right now only SMS_INVITATION_REQUEST is generating android  notifs

refatK commented on app/src/main/java/com/example/bigowlapp/utils/NotificationListenerManager.java:
```diff
@@ -54,4 +66,36 @@ private void handleAuthByPhoneNumberFailure(AuthByPhoneNumberFailure notificatio
+        }
+    }
+
+    private void notificationBuilder(Context context, String channel, Notification notification){
+        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, channel)
+                .setSmallIcon(R.drawable.ic_notification)
+                .setContentTitle(channel)
+                .setStyle(new NotificationCompat.InboxStyle()
+                        .addLine(notification.getMessage()));
+        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context);
+        notificationManager.notify(1, builder.build());
```
To make this slightly better, if possible, add an onClick to this notif that goes to the NotificationActivity. Should be a basic intent page switch thing.

refatK reviewed (changes requested):
Requesting some additional improvmements to the functionality, but these additions should be relatively simple

darkheros12 committed:
Merge branch 'staging' into US-71-Notifications

darkheros12 committed:
Update SignUpPageActivity.java

darkheros12 committed:
PR fixed

I added a functionality that if the user click on the notification, it will lead to the notification page.

refatK commented on app/src/main/java/com/example/bigowlapp/viewModel/SignUpViewModel.java:
```diff
@@ -93,4 +101,33 @@ public String randomNumberStringGenerator() {
+
+    public void createNotificationObject(String senderUid, String phoneNumber){
+
+        Notification newNotification = new Notification();
```
Just need a type on this to give it a title

refatK reviewed (commented):
Just a small issue where the Sms invitation notifs have no title, but it should be an easy fix

darkheros12 committed:
PR fixed

- Remove code smells.
- Add parameters to constructor

darkheros12 committed:
Merge branch 'staging' into US-71-Notifications

ShawnGuerra-Bautista reviewed (approved):
Work perfectly fine, and all of my requested changes have been addressed.

refatK reviewed (approved):
The recent change seems to have fixed the title issue so I  think its fine now
